edge/pkg/edgehub/upgrade/image_prepull.go

```diff
@@ -0,0 +1,195 @@
+	name := message.GetGroup()
+	return name == cloudmodules.ImagePrepullControllerModuleGroup
+}
+
+func (uh *prepullHandler) Process(message *model.Message, clientHub clients.Adapter) error {
+	// get edgecore config
+	edgecoreconfig := options.GetEdgeCoreConfig()
+	if edgecoreconfig.Modules.Edged.TailoredKubeletConfig.ContainerRuntimeEndpoint == "" {
+		edgecoreconfig.Modules.Edged.TailoredKubeletConfig.ContainerRuntimeEndpoint = edgecoreconfig.Modules.Edged.RemoteRuntimeEndpoint
+	}
+	nodeName := edgecoreconfig.Modules.Edged.HostnameOverride
+
+	var errmsg string
+	jobName, err := parseJobName(message.GetResource())
+	if err != nil {
+		errmsg = fmt.Sprintf("failed to parse prepull resource, err: %v", err)
+		sendResponseMessage(nodeName, jobName, errmsg, v1alpha1.PrepullFailed)
+		return fmt.Errorf(errmsg)
+	}
+
+	// parse message request
+	var prepullReq commontypes.ImagePrepullJobRequest
+	data, err := message.GetContentData()
+	if err != nil {
+		errmsg = fmt.Sprintf("failed to get message content, err: %v", err)
+		sendResponseMessage(nodeName, jobName, errmsg, v1alpha1.PrepullFailed)
+		return fmt.Errorf(errmsg)
+	}
+	err = json.Unmarshal(data, &prepullReq)
+	if err != nil {
+		errmsg = fmt.Sprintf("failed to unmarshal message content, err: %v", err)
+		sendResponseMessage(nodeName, jobName, errmsg, v1alpha1.PrepullFailed)
+		return fmt.Errorf(errmsg)
+	}
+	if nodeName != prepullReq.NodeName {
+		errmsg = fmt.Sprintf("request node name %s is not match with the edge node %s", prepullReq.NodeName, nodeName)
+		sendResponseMessage(nodeName, jobName, errmsg, v1alpha1.PrepullFailed)
+		return fmt.Errorf(errmsg)
+	}
+
+	// todo check items such as disk according to req.checkitem config
+
+	// pull images
+	container, err := util.NewContainerRuntime(edgecoreconfig.Modules.Edged.TailoredKubeletConfig.ContainerRuntimeEndpoint, edgecoreconfig.Modules.Edged.TailoredKubeletConfig.CgroupDriver)
+	if err != nil {
+		errmsg = fmt.Sprintf("failed to new container runtime: %v", err)
+		sendResponseMessage(nodeName, jobName, errmsg, v1alpha1.PrepullFailed)
+		return fmt.Errorf(errmsg)
+	}
+
+	go prepullImages(nodeName, jobName, prepullReq, container)
+	return nil
+}
+
+func sendResponseMessage(nodeName, jobName, pullMsg string, state v1alpha1.PrepullState) {
+	resp := commontypes.ImagePrepullJobResponse{
+		NodeName: nodeName,
+		State:    state,
+		Message:  pullMsg,
+	}
+	msg := model.NewMessage("").SetRoute(modules.EdgeHubModuleName, modules.HubGroup).
+		SetResourceOperation(fmt.Sprintf("node/%s/prepull/%s", nodeName, jobName), "prepull").FillBody(resp)
+	beehiveContext.Send(modules.EdgeHubModuleName, *msg)
+}
+
+func prepullImages(nodeName, jobName string, prepullReq commontypes.ImagePrepullJobRequest, container util.ContainerRuntime) {
+	authConfig, err := makeAuthConfig(prepullReq.Secret)
+	if err != nil {
+		errmsg := fmt.Sprintf("failed to get prepull secret, err: %v", err)
+		sendResponseMessage(nodeName, jobName, errmsg, v1alpha1.PrepullFailed)
```

Comment:
Maybe we can decouple the function `sendResponseMessage(...)`
```golang
go func() {
    msg, state := prepullImages(...)
    sendResponseMessage(nodeName, jobName, msg, state)
}()
```

Reply:
> Maybe we can decouple the function `sendResponseMessage(...)`
> 
> ```go
> go func() {
>     msg, state := prepullImages(...)
>     sendResponseMessage(nodeName, jobName, msg, state)
> }()
> ```

done.